src/pygeoprocessing/geoprocessing.py

```diff
@@ -632,6 +648,120 @@ def array_equals_nodata(array, nodata):
     return numpy.isclose(array, nodata, equal_nan=True)
 
 
+def choose_dtype(*raster_paths):
+    """
+    Choose an appropriate dtype for an output derived from the given inputs.
+
+    Returns the dtype with the greatest size/precision among the inputs, so
+    that information will not be lost.
+
+    Args:
+        *raster_paths: series of raster path strings
+
+    Returns:
+        numpy dtype
+    """
+    dtypes = []
+    for path in raster_paths:
+        raster = gdal.OpenEx(path, gdal.OF_RASTER)
+        band = raster.GetRasterBand(1)
+        dtypes.append(_gdal_to_numpy_type(band))
+    return numpy.result_type(*dtypes)
+
+
+def choose_nodata(dtype):
+    """
+    Choose an appropriate nodata value for data of a given dtype.
+
+    Args:
+        dtype (numpy.dtype): data type for which to choose nodata
+
+    Returns:
+        number to use as nodata value
+    """
+    try:
+        return float(numpy.finfo(dtype).max)
+    except ValueError:
+        return int(numpy.iinfo(dtype).max)
+
+
+def raster_map(op, *rasters, target_path='', target_nodata=None,
+              target_dtype=None):
```

Comment:
`target_path=''` is an interesting default value!  What happens if this is not overridden by the user?

Reply:
Related, I wonder: would [keyword-only arguments](https://peps.python.org/pep-3102/#specification) be useful here?  I think that might be the intent with the way these are expressed here, but wanted to check!

Reply:
Good point! Yep, I think I wanted it to be a keyword-only argument. In retrospect, I think it's cleaner to have `rasters` be a list. That more clearly separates what are the input paths vs the output path.

Reply:
Well, I can't decide. Which do you like better?
```
raster_map(lambda a: a * 2, path, target_path=out_path)
raster_map(lambda a, b: a + b, path_a, path_b, target_path=out_path)

or

raster_map(lambda a: a * 2, [path], out_path)
raster_map(lambda a, b: a + b, [path_a, path_b], out_path)
```

Reply:
@phargogh I'm not sure I understand the difference between `raster_map(func, *input_paths, *, target_path)` and `raster_map(func, *input_paths, target_path)` - they seem to work the same? I do like that in this format, you don't have to use a list when passing in just one raster.

Reply:
Uffda, I'm sorry I confused myself here, @emlys !  I had the function signatures wrong in my snippet above.  The corrected options would be 
```python
def raster_map_1(func, input_paths, target_path):
    pass


def raster_map_2(func, *input_paths, target_path):
    pass


raster_map_1(lambda x: x**2, input_paths=['foo.tif', 'bar.tif'], target_path='baz.tif')
raster_map_1(lambda x: x**2, ['foo.tif', 'bar.tif'], 'baz.tif')

# This function can only be called this way; target_path must be a kwarg.
# Input_paths cannot be provided as a kwarg.
raster_map_2(lambda x: x**2, 'foo.tif', 'bar.tif', target_path='baz.tif')
```

Reply:
@emlys and @phargogh, my instinct is to go with option 1 because it feels most at home in the current API design. In another programming language we might be able to easily overload the function based on data type and allow for a single path OR a list of paths easily.

I also find it mildly annoying when entering a single path into a list or even a list of (path, band) tuples in some API instances. But in my opinion if I was using `raster_map` with multiple inputs I'd want to pass a list.

Reply:
I prefer option 1, though mostly from the perspective of someone reading code, rather than writing it. If I read code written with option 2, and I'm not yet super familiar with the function signature, it won't be immediately obvious that `foo.tif` and `bar.tif` are both inputs for the function. Whereas with option 1, there are always 3 args, so it's easy to intuit that the 1st is the function, 2nd is the inputs, 3rd is the output.